extensions/interactions/FractionInput/directives/FractionInput.js

```diff
@@ -12,8 +12,142 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+oppia.directive('oppiaInteractiveFractionInput', [
+  'UrlInterpolationService',
+  function(UrlInterpolationService) {
+    return {
+      restrict: 'E',
+      scope: {
+        onSubmit: '&'
+      },
+      templateUrl: UrlInterpolationService.getExtensionResourceUrl(
+        '/interactions/FractionInput/directives/' +
+        'fraction_input_interaction_directive.html'),
+      controller: [
+        '$scope', '$attrs', 'FocusManagerService',
+        'fractionInputRulesService',
+        'FractionObjectFactory', 'FRACTION_PARSING_ERRORS',
+        function($scope, $attrs, FocusManagerService,
+          fractionInputRulesService,
+          FractionObjectFactory, FRACTION_PARSING_ERRORS) {
+          $scope.answer = '';
+          $scope.labelForFocusTarget = $attrs.labelForFocusTarget || null;
+          var requireSimplestForm =
+            $attrs.requireSimplestFormWithValue === 'true';
+          var errorMessage = '';
+          var FORM_ERROR_TYPE = 'FRACTION_FORMAT_ERROR';
+          $scope.FRACTION_INPUT_FORM_SCHEMA = {
+            type: 'unicode',
+            ui_config: {}
+          };
+
+          $scope.getWarningText = function() {
+            return errorMessage;
+          };
+
+          /**
+           * Disables the input box if the data entered is not a valid prefix
+           * for a fraction.
+           * Examples of valid prefixes:
+           * -- 1
+           * -- 1 2
+           * -- 1 2/
+           * -- 2/
+           * -- 1 2/3
+           */
+          $scope.$watch('answer', function(newValue) {
+            var INVALID_CHARS_REGEX = /[^\d\s\/-]/g;
+            // Accepts incomplete fraction inputs
+            // (see examples above except last).
+            var PARTIAL_FRACTION_REGEX =
+              /^\s*(-?\s*((\d*\s*\d+\s*\/?\s*)|\d+)\s*)?$/;
+            // Accepts complete fraction inputs.
+            var FRACTION_REGEX =
+              /^\s*-?\s*((\d*\s*\d+\s*\/\s*\d+)|\d+)\s*$/;
+            if (INVALID_CHARS_REGEX.test(newValue)) {
+              errorMessage = FRACTION_PARSING_ERRORS.INVALID_CHARS;
+              $scope.FractionInputForm.answer.$setValidity(
+                FORM_ERROR_TYPE, false);
+            } else if (!(FRACTION_REGEX.test(newValue) ||
+                PARTIAL_FRACTION_REGEX.test(newValue))) {
+              errorMessage = FRACTION_PARSING_ERRORS.INVALID_FORMAT;
+              $scope.FractionInputForm.answer.$setValidity(
+                FORM_ERROR_TYPE, false);
+            } else {
+              errorMessage = '';
+              $scope.FractionInputForm.answer.$setValidity(
+                FORM_ERROR_TYPE, true);
+            }
+          });
+
+          $scope.submitAnswer = function(answer) {
+            try {
+              var fraction = FractionObjectFactory.fromRawInputString(
```

Comment:
I left it within the try block because it wouldn't make sense to process the following code if an error occurs (i.e. submitting the fraction which would be undefined). Is that okay or is there a better way to handle it?

Reply:
Ah, yes, I missed that you weren't throwing again in the catch(). OK, I think this is fine.